django/db/models/fields/related.py

```diff
@@ -1252,11 +1252,18 @@ def set_managed(model, related, through):
     lazy_related_operation(set_managed, klass, to_model, name)
 
     to = make_model_tuple(to_model)[1]
+    to_app = make_model_tuple(to_model)[0]
     from_ = klass._meta.model_name
-    if to == from_:
+    has_conflicting_m2m_field = False
+
+    if to == from_ and to_app == klass._meta.app_label:
         to = "to_%s" % to
         from_ = "from_%s" % from_
 
+    if to == from_ and to_app != klass._meta.app_label:
+        to = "%s_%s" % (to_app, to)
+        has_conflicting_m2m_field = True
+
```

Comment:
I tried many cases in a sample project and it was working fine for me. The tables that were created before we attempt to change the naming conventions were recreated (for sqlite3) without any data loss. here is the sql for the same:
```
BEGIN;
--
-- Rename model Folder to Incident
--
ALTER TABLE "vault_folder" RENAME TO "vault_incident";
CREATE TABLE "vault_incident_incidents" ("id" integer NOT NULL PRIMARY KEY AUTOINCREMENT, "incident_id" bigint NOT NULL REFERENCES "vault_incident" ("id") DEFERRABLE INITIALLY DEFERRED, "incidents_incident_id" bigint NOT NU
LL REFERENCES "incidents_incident" ("id") DEFERRABLE INITIALLY DEFERRED);
INSERT INTO "vault_incident_incidents" (id, incident_id, incidents_incident_id) SELECT id, folder_id, incident_id FROM "vault_folder_incidents";
DROP TABLE "vault_folder_incidents";
CREATE UNIQUE INDEX "vault_incident_incidents_incident_id_incidents_incident_id_3bd83bac_uniq" ON "vault_incident_incidents" ("incident_id", "incidents_incident_id");
CREATE INDEX "vault_incident_incidents_incident_id_ef11b7b1" ON "vault_incident_incidents" ("incident_id");
CREATE INDEX "vault_incident_incidents_incidents_incident_id_9cababa3" ON "vault_incident_incidents" ("incidents_incident_id");
COMMIT;
```

Reply:
@DevilsAutumn you thing I believe you didn't account for in your original solution is the case where a `app1.Model.ManyToMany(app2.Model)` exists already against a past version of Django.

With the code in `main` the intermediary table columns would have been `from_model_id`, `to_model_id` but your proposed changes would have made the column names expected by the ORM to be `model_id`, `app2_model_id`

Reply:
> With the code in `main` the intermediary table columns would have been `from_model_id`, `to_model_id` but your proposed changes would have made the column names expected by the ORM to be `model_id`, `app2_model_id`

yes ,exactly! initially i thought this naming convention would be easier to understand in case of model name conflicts in different apps as ```from``` and ```to``` are used for self referential m2m field but here that is not the case. So, to avoid that confusion i tried a diff naming convention.